FuelLabs/fuel-indexer
# fix: --embedded-database does not work when using fuel-indexer binary directly

### Description

This PR makes `fuel-indexer start` run an embedded `postgres` database when `--embedded-database` flag is specified.

When `fuel-indexer` exits, the embedded database is shut down. For this purpose, `fuel-indexer` now listens to `SIGHUP/INT/TERM` signals.

### Testing steps

`fuel-indexer`

1. `--embedded-database`
```
cargo run -p fuel-indexer -- run --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80 --auth-enabled --auth-strategy jwt --jwt-secret $FUEL_JWT_TOKEN --jwt-issuer FuelLabs --postgres-user postgres --postgres-password postgres --postgres-database postgres9 --postgres-port 12345 --embedded-database
```

2. `--embedded-database` and `--local-fuel-node`
```
cargo run -p fuel-indexer --features api-server,fuel-core-lib -- run --run-migrations --postgres-user postgres --postgres-password postgres --postgres-database postgres9 --postgres-port 12345 --embedded-database --local-fuel-node
```
Note: `fuel-core-lib` must be enabled.

3 `--embedded-database` and `--all-features`, `--no-default-features`

```
cargo run -p fuel-indexer --all-features -- run --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80 --postgres-user postgres --postgres-password postgres --postgres-database postgres9 --postgres-port 12345 --embedded-database
```

```
cargo run -p fuel-indexer --no-default-features -- run --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80 --postgres-user postgres --postgres-password postgres --postgres-database postgres9 --postgres-port 12345 --embedded-database
```

4. `fuel-api-server` `--all-features`, `--no-default-features`

```
cargo run -p fuel-indexer-api-server --all-features -- run --fuel-node-host beta-3.fuel.network --fuel-node-port 80
```

```
cargo run -p fuel-indexer-api-server --no-default-features -- run --fuel-node-host beta-3.fuel.network --fuel-node-port 80
```

And check it works by visiting:
http://localhost:29987/api/playground/fuel_examples/explorer_indexer

### Changelog

* start an embedded database when `--embedded-database` is specified
* listen for termination signals
* stop the embedded database when a termination signal is received
* cleanup: `forc index start` passes `--embedded-database` flag to `fuel-indexer` instead of calling `forc_postgres::commands::create::exec`
* change how handles are created for various features—no nested cfg's
* vendored `openssl`
    it's a fix for
    ```
    ./fuel-node: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
    ```
    which was suggested here:
    https://github.com/launchbadge/sqlx/issues/473
    "the issue is because procedural macro executed in a slightly different environment"
    I think we're running into the same situation with our Docker build—we build with `cargo chef` image, but run on `ubuntu:22.04`
